sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/transformwithstate/testing/InMemoryStatefulProcessorHandle.scala

```diff
@@ -0,0 +1,332 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.streaming.operators.stateful.transformwithstate.testing
+
+import java.time.{Clock, Instant}
+import java.util.UUID
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+import org.apache.spark.sql.Encoder
+import org.apache.spark.sql.execution.streaming.operators.stateful.transformwithstate.statefulprocessor.ImplicitGroupingKeyTracker
+import org.apache.spark.sql.execution.streaming.operators.stateful.transformwithstate.statefulprocessor.QueryInfoImpl
+import org.apache.spark.sql.streaming.{ListState, MapState, QueryInfo, StatefulProcessorHandle, TimeMode, TTLConfig, ValueState}
+
+/** Helper to track expired keys. */
+class TtlTracker(val clock: Clock, ttl: TTLConfig) {
+  require(!ttl.ttlDuration.isNegative())
+  private val keyToLastUpdatedTime = mutable.Map[Any, Instant]()
+
+  def isKeyExpired(): Boolean = {
+    if (ttl.ttlDuration.isZero()) {
+      return false
+    }
+    val key = ImplicitGroupingKeyTracker.getImplicitKeyOption.get
+    if (!keyToLastUpdatedTime.contains(key)) {
+      return false
+    }
+    val expiration: Instant = keyToLastUpdatedTime.get(key).get.plus(ttl.ttlDuration)
+    return expiration.isBefore(clock.instant())
+  }
+
+  def onKeyUpdated(): Unit = {
+    val key = ImplicitGroupingKeyTracker.getImplicitKeyOption.get
+    keyToLastUpdatedTime.put(key, clock.instant())
+  }
+}
+
+/** In-memory implementation of ValueState. */
+class InMemoryValueState[T](clock: Clock, ttl: TTLConfig) extends ValueState[T] {
+  private val keyToStateValue = mutable.Map[Any, T]()
+  private val ttlTracker = new TtlTracker(clock, ttl)
+
+  private def getValue: Option[T] = {
+    if (ttlTracker.isKeyExpired()) {
+      return None
+    }
+    keyToStateValue.get(ImplicitGroupingKeyTracker.getImplicitKeyOption.get)
+  }
+  override def exists(): Boolean = getValue.isDefined
+
+  override def get(): T = getValue.getOrElse(null.asInstanceOf[T])
+
+  override def update(newState: T): Unit = {
+    ttlTracker.onKeyUpdated()
+    keyToStateValue.put(ImplicitGroupingKeyTracker.getImplicitKeyOption.get, newState)
+  }
+
+  override def clear(): Unit =
+    keyToStateValue.remove(ImplicitGroupingKeyTracker.getImplicitKeyOption.get)
+}
+
+/** In-memory implementation of ListState. */
+class InMemoryListState[T](clock: Clock, ttl: TTLConfig) extends ListState[T] {
+  private val keyToStateValue = mutable.Map[Any, mutable.ArrayBuffer[T]]()
+  private val ttlTracker = new TtlTracker(clock, ttl)
+
+  private def getList: Option[mutable.ArrayBuffer[T]] = {
+    if (ttlTracker.isKeyExpired()) {
+      return None
+    }
+    keyToStateValue.get(ImplicitGroupingKeyTracker.getImplicitKeyOption.get)
+  }
+
+  override def exists(): Boolean = getList.isDefined
+
+  override def get(): Iterator[T] = {
+    getList.orElse(Some(mutable.ArrayBuffer.empty[T])).get.iterator
```

Comment:
```suggestion
    getList.getOrElse(mutable.ArrayBuffer.empty[T]).iterator
```

I guess this would work?